### DISCUSSION TITLE
NOT IN (x, x, x) filter

### DISCUSSION THREAD

**Discussion by rgr** [meilisearch/product#274] [Feedback & Feature Proposal]
Hi there,

it would be great to be able to filter the search while removing ids of resources for external jointure reasons. I see we can do it one by one in the offered filters but, is there a way to do it for a large array of resources ? That would be interesting.

Thanks !

**Comment by gmourier** [comment]
Hi @rgr 👋,

If I understand correctly rather than doing a filter `id = 1 AND id = 2 AND id = 5678`. Would you like an IN operator to write `id IN (1, 2, 5678)` directly?

I'm not sure to understand that particular part "while removing ids of resources for external jointure reasons". Can you describe your use case in more detail?

Thanks for the feedback and your answer!

**Reply by rgr** [reply]
Well yeah I was thinking about something like that but in the end I had this working (although I don't know how well it will handle thousands of data points) :

```php
$options['filter'] = $options['filter'] . ' AND id!=' . implode(' AND id!=', $userId);
```

**Comment by gmourier** [comment]
Hello dear community, 👋

We have published a docker image allowing you to test the addition of the `IN` operator.

### `IN` operator

Instead of writing a cumbersome clause like `user_id = 1 OR user_id = 2 OR user_id = 3 ...`. It's possible to write `user_id IN [1, 2, 3]` and voila!

- `user_id IN [...]` selects all documents whose attribute contains at least one of the values given inside the brackets. In case there are no values, then the filter selects no documents.
- `NOT user_id IN[...]` selects all documents whose attribute contains values that are all different from those in the brackets. (It includes documents missing the attribute - See `EXISTS` operator for more granularity).

---

We have also made some other changes that may be of interest to you.

### `NOT` / `!=` behavior

`NOT` /`!=` operator has been enhanced following this [discussion](https://github.com/meilisearch/product/discussions/325).

`tags != adult` now returns documents having `tags: null` or `tags: []` but also documents without the `tags` property. It's equivalent to `NOT tags = adult`.

The `NOT` operator selects all documents not selected by the filter that follows it.

### `EXISTS` operator

We took the opportunity to add the operator `EXISTS`. This operator allows you to granularly require that the selected documents have a `tag` attribute regardless of its value.

In this case, `tags != adult and tags EXISTS` to retrieve documents that do not contain the `adult` value in the `tag` attribute and only those for which the `tag` attribute exists.

- `tags EXISTS` selects all documents having the `tags` attribute.
- `NOT tags EXISTS` selects all documents missing the `tags` attribute.

---

We would like to have your feedback before deciding whether to introduce the given changes in a future version.

To do so you have to pull the docker image we prepared.

```bash
docker run -it --rm \
    -p 7700:7700 \
    -v $(pwd)/meili_data:/meili_data \
    getmeili/meilisearch:v0.29.0-filter.beta.0
```

Thank you so much for being part of our community ✨

**Comment by gmourier** [comment]
Hello 👋

We have just released [v0.29.0rc1](https://github.com/meilisearch/meilisearch/releases/tag/v0.29.0rc1), which is a release candidate of v0.29.0 🔥

You can test the just-introduced `EXISTS` and `IN` filtering operators!
More information in [this issue](https://github.com/meilisearch/meilisearch/issues/2479) and [this one](https://github.com/meilisearch/meilisearch/issues/2484).

[Binaries](https://github.com/meilisearch/meilisearch/releases/tag/v0.29.0rc1) are attached to the release, or you can use the docker image:
```bash
docker run -it --rm \
    -p 7700:7700 \
    getmeili/meilisearch:v0.29.0rc1
```

Let us know about any bugs or feedback! 😄 It would be really helpful.

FYI, the official v0.29.0 release will be available on 3rd October.

**Comment by gmourier** [comment]
Hey everyone 👋

Locking this discussion since the `IN` operator has been released.

`IN [valueA, valueB, …, valueN]` selects all documents whose chosen field contains at least one of the specified values. The expression `genres IN [horror, comedy]` returns all documents whose genres includes either horror, comedy, or both

📚 https://docs.meilisearch.com/learn/advanced/filtering_and_faceted_search.html#using-filters

Please open a new discussion if you have any feedback to share 🙇‍♂️
